Tracing docker buildx commands

twunderlich-grapl opened the pull request:
<!-- Thank you for your contribution to Grapl! Please take some time
to fill out this short template to help us review and understand your
PR. -->

### Which issue does this PR correspond to?
https://github.com/grapl-security/issue-tracker/issues/862
<!-- Please provide a link to the GitHub Issue this pull request is
meant to address. -->

### What changes does this PR make to Grapl? Why?
This adds commands to enable/disable tracing for docker buildx bake
<!-- Please describe the functional impact of these changes and the
rationale behind them. This will help us understand your approach to
solving the problem. -->

### How were these changes tested?
Run `make up` (so you have Jaeger locally)
Run `make buildx-tracing`
Run `make build-local-infrastructure` or other build command as long as it uses buildx
Go to http://localhost:16686/ and look at the buildkite traces.
Run `make disable-buildx-tracing`
Run the build again 
Refresh Jaeger. No new traces will be there
<!-- Please describe how these changes were tested. There should be
sufficient detail that reviewers can replicate your testing
procedure. If the procedure is a manual one that's OK, your
description will help us work with you to get it automated. -->

Commit:
Tracing docker buildx commands

This makes it easier to diagnose why docker builds are slow.

wimax-grapl commented on Makefile:
```diff
@@ -633,6 +633,19 @@ generate-nomad-rust-client: ## Generate the Nomad rust client from OpenAPI
 # should use that here.
 	$(MAKE) format-prettier
 
```
1. Could we get a quick README or something on how to use and consume this?
2. Not necessarily for _this_ PR but would it make sense to exercise this in CI as a "hey here's proof the jaeger setup all still works"?

wimax-grapl replied:
Also, in general, it seems odd to me that this is a
1. enable
2. do things
3. disable

and not a 
1. WITH_JAEGER_TRACING=1 do things

wish we had context managers or RAII in bash...

christophermaier replied:
I had a similar thought... I wonder if we could have a one-time "setup" target that creates the tracing-enabled builder, and then a `WITH_TRACING` macro / flag / etc. that would add the `--builder=XXX` option to the `docker buildx bake` invocations.

``` make
ifdef WITH_TRACING
buildx_builder_args := --builder=grapl-tracing
endif

DOCKER_BUILDX_BAKE_HCL := docker buildx bake --file=docker-bake.hcl $(buildx_builder_args)
```
Then you can do `WITH_TRACING=1 make local-infrastructure` to get tracing enabled (given the builder exists, of course).

(If there was anything else to add tracing to, we could stick that behind the `WITH_TRACING` variable, too.)

It's still not airtight, though, because it still relies on a) having the builder set up and b) that Jaeger is running.

Make is kinda weird for this 🤷 

(These are just thoughts... either way is fine; I'm thinking out loud here)

christophermaier replied:
Taking another look, I think the existing code conflates the _creation_ of a builder with the _using_ of that builder. In particular, disabling the builder does not also delete that builder (I don't think that it _should_ do that, just to be clear), so I think we'll have an error the next time we run `make buildx-tracing` because it will try to create another builder with the same name as one that already exists.

Decoupling the creation of the builder from the using or not-using of that builder seems like a worthwhile change.

twunderlich-grapl replied:
1. I've added a README in the docs/development folder
2. I've updated the code to use an environment variable and a one-time setup command

wimax-grapl committed:
Soft-fail rootfs builds in CI until we figure out what's going on (#1546)

* Soft-fail rootfs builds in CI until we figure out what's going on

* english is hard

christophermaier commented on Makefile:
```diff
@@ -633,6 +633,19 @@ generate-nomad-rust-client: ## Generate the Nomad rust client from OpenAPI
+.PHONY: buildx-tracing
+buildx-tracing: ## Set tracing up for docker buildx, this assumes that Jaeger is already up
+	docker buildx create \
+      --name builder \
```
It'd be good to name this something more descriptive, like "grapl-tracing", or "grapl-jaeger", etc.

twunderlich-grapl replied:
So since this only sets tracing for docker (we set up tracing for the grapl services without this), I think it makes more sense to rename this docker-tracing than grapl-tracing.

christophermaier replied:
I meant that, among the _n_ potential builders you might one day have on your system, this would be the one that Grapl has set up for tracing builds. It's inherently for Docker since it's a Docker builder 🙃

twunderlich-grapl replied:
Ah, you meant the actual builder name, not the make command 🤦 . Fixing now

wimax-grapl committed:
Resolve apt failures with a 'cloud-init status --wait', change to server-minimal us-east-2 AMI (#1548)

* Resolve apt failures with a 'cloud-init status --wait', change to server-minimal AMI, change to us-east-2 AMI

* ami-id to ami_id woops

* Revert makefile change

* Back to us-east-1 due to us-east-2 no default vpc; not worth debugging

wimax-grapl committed:
Add a `make dist/plugin-bootstrap-init`  (#1542)

* Add lots of #s to the rust dockerfile to make each stage stand out more

* Add a 'make dist/plugin-bootstrap-init', which will then be consumed by rootfs

* format

* Add lots of documentation about the PHONYness of dist/plugin-bootstrap-init and its downstream dist/firecracker_rootfs

* Slightly refactor the way files in plugin-bootstrap-init-output get spit out to dist/plugin-bootstrap-init/ (move directory to bake.hcl)

* Clarify weirdness of COPYing source stuff into dist/

wimax-grapl committed:
Remove unused services from Localstack (#1550)

* Remove unused services from Localstack

* format

Commit:
Make docker traces more ergonomical

This makes enabling docker buildx traces much easier. We now have a
global setup step, and then one just needs an env var. This is more in
line with our current patterns

christophermaier commented on Makefile:
```diff
@@ -29,11 +29,16 @@ export
 
+# This is used to send docker traces to Jaeger. This is primarily useful for debugging build time performance
+ifdef WITH_TRACING
+buildx_builder_args := --builder=builder --load
```
Doesn't `--load` output a tarball? 

https://github.com/moby/buildkit#docker-tarball

christophermaier replied:
🤔 Hrmm... maybe not; I don't see that it does anything different 🤔

Commit:
Remove the --load parameter for docker build

The load parameter outputs a tarball. When using the docker-container
driver this is required for the built images to show up in `docker
image ls`, however it does slow down the builds. Since tracing docker
builds will be used primarily for debugging docker problems and build
performance, this seems like a useful optimization

Commit:
Add docker buildx tracing to other services

This adds tracing for other services like tracing and formatters

Commit:
Add docs for tracing usage

This adds general tracing instructions and docker build tracing
instructions

Commit:
Rename docker driver name to be explicit

We're renaming this so that if someone has multiple docker drivers
installed and potentially used for other projects they will know that
this docker driver is the one to use for tracing docker builds in the
grapl project

Commit:
Merge branch 'main' into trace-docker-build

 Conflicts:
	nomad/local/grapl-local-infra.nomad

Commit:
Make build-docs happy
